Add Time Series Block

codeboy5 opened the pull request:
Added Time Series Container and Block. It is capable of loading all datasets from [timeseriesclassification](http://timeseriesclassification.com/dataset.php). The `.ts` files are loaded using Julia translation of this [method](https://github.com/timeseriesAI/tsai/blob/600a5ee7efe04c6ff19c7a27f596f947c75f31ef/tsai/data/external.py#L1389) .

I have also added a basic test case for the recipe. 
This allows us to do the following 


```julia
using FastAI

data, blocks = load(datarecipes()["ecg5000"])
nobs(data)
sample = series, class = getobs(data, 10)
```

Just wanted to get some initial thoughts on the work, there might be more changes as I continue to work on the other parts.

codeboy5 commented:
Hey for downloading the datasets, the message as well as the post fetch method needs to be different which I think might require some restructuring, how should I go about that ?

ToucheSir commented on src/datasets/load.jl:
```diff
@@ -1,42 +1,42 @@
 If it hasn't been downloaded yet, you will be asked if you want to
 download it. See [`Datasets.DATASETS`](#) for a list of available datasets.
 """
 function datasetpath(name)
```
Looks like the indentation got a little messed up in this file?

codeboy5 replied:
Just to clarify, is there a way to auto format the code ?

ToucheSir commented on src/datasets/load.jl:
```diff
@@ -47,17 +47,18 @@ const RE_IMAGEFILE = r".*\.(gif|jpe?g|tiff?|png|webp|bmp)$"i
 const RE_TEXTFILE = r".*\.(txt|csv|json|md|html?|xml|yaml|toml)$"i
 istextfile(f) = matches(RE_TEXTFILE, f)
-
+const RE_TIMESERIESFILE = r".*\.(ts)$"i
```
We'll have to figure out how to deal with time series datasets in different formats eventually. This may involve changing the text file detection logic. No need to worry about that now, this is just a note for when that time comes :)

ToucheSir commented on src/datasets/containers.jl:
```diff
@@ -94,6 +308,21 @@ LearnBase.getobs(dataset::TableDataset{<:CSV.File}, idx) = dataset.table[idx]
 
+#TimeSeriesDataset
+
+struct TimeSeriesDataset{}
+    table::AbstractArray{Float64,3}
```
```suggestion
struct TimeSeriesDataset{T<:AbstractArray}
    table::T
```
The type parameter is required for avoid an abstract field type. I've loosened the array type bound as well, but it may need to be tuned up or down later.

ToucheSir commented on src/datasets/containers.jl:
```diff
@@ -26,6 +26,218 @@ function rglob(filepattern = "*", dir = pwd(), depth = 4)
+                    has_problem_name_tag = true
+                    metadata_started = true
+
+                elseif startswith(ln, "@timestamps")
+                    # Check that the associated value is valid
+                    tokens = split(ln, " ")
+                    token_len = length(tokens)
+
+                    if tokens[2] == "true"
+                        timestamps = true
+                    else
+                        timestamps = false
+                    end
+
+                    has_timestamps_tag = true
+                    metadata_started = true
+
```
There should be a way to extract out some of this logic. For now, having tests would help verify it's working.

codeboy5 replied:
Yeah I think it could be extracted out for the current datasets we are using. I just copied it over from python library to get it working. 
By tests do you mean automated tests and creating a sample `.ts` file to run the test on ?

ToucheSir replied:
Correct. You could also use a small existing one as long as the license is compatible.

codeboy5 commented on src/TimeSeries/blocks/timeseriesrow.jl:
```diff
@@ -0,0 +1,37 @@
+
+"""
+
+struct TimeSeriesRow{M,N} <: Block end
```
- For the current datasets we are planning to use, the length is same for all observations, so we can have a constant block.

From my understanding we build the model from the block ? So the parameters might depend on time series length and number of variables. 
So ideally the block should just dispatch on parameters which are required in model building ?

codeboy5 commented on src/TimeSeries/recipes.jl:
```diff
@@ -0,0 +1,63 @@
+# ## Tests
+
+@testset "TimeSeriesDataset [recipe]" begin
+    path = datasetpath("ecg5000")
```
The dataset is around 10mb, should we run this on CI ?

lorenzoh replied:
That should be fine 👍

codeboy5 commented on src/datasets/containers.jl:
```diff
@@ -94,6 +308,21 @@ LearnBase.getobs(dataset::TableDataset{<:CSV.File}, idx) = dataset.table[idx]
+    table::AbstractArray{Float64,3}
+end 
+
+function LearnBase.getobs(dataset::TimeSeriesDataset, idx)
```
Sure, will update it.

codeboy5 commented:
I have implemented the changes we discussed in the last call.

- [Replace type parameters with fields](https://github.com/FluxML/FastAI.jl/blob/9117ac6de2ceaa4d590186d3e990c1f0dd803419/src/TimeSeries/blocks/timeseriesrow.jl#L26)
- [TSClassificationDatasetStruct](https://github.com/FluxML/FastAI.jl/blob/9117ac6de2ceaa4d590186d3e990c1f0dd803419/src/datasets/fastaidatasets.jl#L12) to load the `.ts` files from http://timeseriesclassification.com .
- Test for `.ts2df` function as @ToucheSir suggested. The size for file is 159KB, should I just include it in the commit or download It on the fly. As for the license, according to [tsml_repo](https://github.com/uea-machine-learning/tsml_repo), it seems to have a [GNU General Public License v3.0](https://github.com/uea-machine-learning/tsml_repo/blob/master/LICENSE) .

ToucheSir commented:
Great. Just to be safe, I think it would be best to download the file on the fly for now.

lorenzoh commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,51 @@
+    return mean, std
+end
+
+function setup(::Type{TSPreprocessing}, ::TimeSeriesRow, data)
+    means, stds = tsdatasetstats(data)
+end
```
```suggestion
setup(::Type{TSPreprocessing}, ::TimeSeriesRow, data) = means, stds = tsdatasetstats(data)
```

lorenzoh commented on src/TimeSeries/TimeSeries.jl:
```diff
@@ -31,12 +31,22 @@ using InlineTest
 
-export TimeSeriesRow
-
+export 
+    TimeSeriesRow, TSClassificationSingle, TimeSeriesPreprocessing
```
```suggestion
export 
    TimeSeriesRow, TSClassificationSingle, TSPreprocessing
```

lorenzoh commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,51 @@
+    tfms
+end
+
+function TSPreprocessing()
+    base_tfms = [
+    ]
+    return TSPreprocessing(base_tfms)
+end
```
What kind of transforms will be in here?

lorenzoh commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,51 @@
+    return TSPreprocessing(base_tfms)
+end
+
+function encodedblock(p::TSPreprocessing, block::TimeSeriesRow)
+    return block
+end
```
If the format of the time series is changed by the encoding, this should return a different block

codeboy5 replied:
No the format won't be changed, as I discussed with Brian earlier that different models might require different formats and so the encoding shouldn't depend on the model.

codeboy5 commented:
I have added the basic structure and code. There seems to be a couple of errors, I will solve by tonight. 
Let me know how does the basic code look.

lorenzoh commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -1,3 +1,12 @@
+struct TSStats{T}
+    means::AbstractArray{T,2}
+    stds::AbstractArray{T,2}
+end
+
+function TSStats(means, stds)
+    TSStats{eltype(means)}(means, stds)
+end
```
Not sure we need this struct, it may be simpler to add `means` and `stds` fields to the `Encoding`. Then that also makes it easier to construct `TSPreprocessing` manually (i.e. without `setup`).

lorenzoh commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -10,23 +19,25 @@ Encodes
-        obs = tfm(obs)
-    end
-    obs
+function TSStandardize(
```
Should this be an `encode` method?

lorenzoh commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -41,11 +52,26 @@ function tsdatasetstats(
+    return block
+end
+
+function encode(p::TSPreprocessing, context, block::TimeSeriesRow, obs)
```
Can we add a check of the dimensions of `obs` against the means/stds here?

lorenzoh reviewed (commented):
Left some minor comments, but let me know when you're done with the other changes :) :+1:

ToucheSir commented on src/TimeSeries/blocks/timeseriesrow.jl:
```diff
@@ -0,0 +1,42 @@
+    obslength::Union{Int, Colon}
+end
+
+function checkblock(row::TimeSeriesRow, obs::AbstractArray{T,2}) where {T<:Number}
+    size(obs) == (row.nfeatures, row.obslength)
+end
```
```suggestion
checkblock(row::TimeSeriesRow, obs::AbstractMatrix{<:Number}) = size(obs) == (row.nfeatures, row.obslength)
```

codeboy5 replied:
Yeah this looks better, thanks 👍

ToucheSir commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,77 @@
+    obs,
+    p::TSPreprocessing
+)
+    means = p.stats.means
+    stds  = p.stats.stds
+    obs = obs .- means
+    obs = obs ./ stds
+    return obs
```
```suggestion
    return (obs .- p.stats.means) ./ p.stats.stds
```
This will fuse and save an intermediate allocation.

ToucheSir commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,77 @@
+    by_var=false,
+    by_step=false
+)
+    drop_axes = []
+    if (by_var)
+        append!(drop_axes,2)
+    else
+        append!(drop_axes,3)
+    end 
+    axes = [ax for ax in [1, 2, 3] if !(ax in drop_axes)]
```
```suggestion
    drop_axis = by_var ? 2 : 3
    axes = filter(∉(drop_axis), 1:3)
```
It appears `drop_axes` only ever has one axis, so I've taken the liberty to simplify it here. Unless you were planning on having an if statement for `by_step` as well?

codeboy5 replied:
Yeah this would work, we can have the normalisation over sample too but yeah for now this would look better.

ToucheSir commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,77 @@
+    end 
+    axes = [ax for ax in [1, 2, 3] if !(ax in drop_axes)]
+    mean = Statistics.mean(data.table, dims=axes)
+    std  = Statistics.std(data.table, dims=axes)
```
```suggestion
    std  = Statistics.std(data.table, dims=axes, mean=mean)
```
If the data source supports it, this is more efficient.

codeboy5 replied:
Yes, I tried using `Statistics.stdm(itr, mean)` earlier, couldn't seem to get it working. Will look into it further and keep this comment open till then.

ToucheSir commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,77 @@
+
+function setup(::Type{TSPreprocessing}, ::TimeSeriesRow, data)
+    means, stds = tsdatasetstats(data[1])
+    means = reshape(means, ( size( means)[2:3] ))
+    stds  = reshape(stds, ( size( stds)[2:3] ))
```
A little weird formatting going on here. Why is the reshape needed?

ToucheSir commented on src/datasets/containers.jl:
```diff
@@ -1,4 +1,216 @@
+        end
+
+        # Check if we should return any associated class labels separately
+        if class_labels
+            return data, class_val_list
+        else
+            return data
+        end
```
Returning one or two things based on a conditional is a little surprising. Consider either always returning class_val_list and/or making the second return value some meaningful null value (nothing, empty array, whatever makes the most sense).

ToucheSir commented on src/TimeSeries/encodings/timeseriespreprocessing.jl:
```diff
@@ -0,0 +1,77 @@
+struct TSStats{T}
+    means::AbstractArray{T,2}
+    stds::AbstractArray{T,2}
```
You could also consider changing these to `AbstractMatrix{T}`, but it would be best to confirm first that they will always be 2-dimensional (what about higher-dimensional time series, for example?)

codeboy5 replied:
Not sure about the higher dimensional series, I will have to check the literature or some examples online if I can find them.

ToucheSir replied:
Oh I can come up with examples easily enough, the question is whether the fastai docs have any :)

codeboy5 replied:
Oh, I will check that up before our meeting today.

codeboy5 replied:
Oh, I will check that up before our meeting today.

codeboy5 commented:
Hey the encodings are working now.

```julia
> using FastAI
> data, blocks = load(datarecipes()["ecg5000"]);
> task = FastAI.TSClassificationSingle(blocks, data);
> input, target = getobs(data, 2)
(Float32[-1.1008778 -3.9968398 … 1.1196209 -1.4362499], "1")
> encodesample(task, Training(), (input, target))
(Float32[-1.1048299 -4.011188 … 1.1236403 -1.4414059], Float32[1.0, 0.0, 0.0, 0.0, 0.0])
```

I am looking to add some tests as well look to resolve all the comments you guys made before the meeting. Have also started working on creating the demo notebook side-by-side as you guys suggested.

codeboy5 commented:
I've added the decode method and basic visualisation too and updated the notebook accordingly. The formatting for plots Is a little weird on the notebooks, but looks fine on the terminal. Do you guys know how to fix this, should I try adjusting the size of the plot ?

<img width="427" alt="Screenshot 2022-07-06 at 10 07 39 AM" src="https://user-images.githubusercontent.com/40931412/177470543-7c4556c8-819c-42a6-929d-a69e7c1788ea.png">

lorenzoh commented:
> The formatting for plots Is a little weird on the notebooks, but looks fine on the terminal. 

Rendering of Unicode plots can be a bit off depending on the notebook environment/viewer. It should end up looking fine in the docs, though.

lorenzoh commented:
Having some trouble to push the subpackage merge to your fork, @codeboy5. Can you check that I have permissions to push to your fork? Or, alternatively just give me committer access to the fork

![image](https://user-images.githubusercontent.com/28812146/177519557-7a960778-d47e-4851-8d21-41bbc9aedcaf.png)

codeboy5 commented:
> Having some trouble to push the subpackage merge to your fork, @codeboy5. Can you check that I have permissions to push to your fork? Or, alternatively just give me committer access to the fork
> 
> ![image](https://user-images.githubusercontent.com/28812146/177519557-7a960778-d47e-4851-8d21-41bbc9aedcaf.png)

If checked, users with write access to FluxML/FastAI.jl can add new commits to your timeseries-blocks branch. You can always change this setting later.
This option is checked out for me.
Should I add you as a collaborator on my forked repo ?

lorenzoh commented:
> > Having some trouble to push the subpackage merge to your fork, @codeboy5. Can you check that I have permissions to push to your fork? Or, alternatively just give me committer access to the fork
> > 
> > ![image](https://user-images.githubusercontent.com/28812146/177519557-7a960778-d47e-4851-8d21-41bbc9aedcaf.png)
> 
> If checked, users with write access to FluxML/FastAI.jl can add new commits to your timeseries-blocks branch. You can always change this setting later.
> This option is checked out for me.
> Should I add you as a collaborator on my forked repo ?

Yeah, please try that

codeboy5 commented:
> > > Having some trouble to push the subpackage merge to your fork, @codeboy5. Can you check that I have permissions to push to your fork? Or, alternatively just give me committer access to the fork
> > > ![image](https://user-images.githubusercontent.com/28812146/177519557-7a960778-d47e-4851-8d21-41bbc9aedcaf.png)
> > 
> > 
> > If checked, users with write access to FluxML/FastAI.jl can add new commits to your timeseries-blocks branch. You can always change this setting later.
> > This option is checked out for me.
> > Should I add you as a collaborator on my forked repo ?
> 
> Yeah, please try that

Yeah, just sent you an invite for the same. Hopefully it should work now.

codeboy5 commented:
With this commit, we are able to the following now 

```julia
> using FastAI, FastTimeSeries
> data, blocks = load(datarecipes()["ecg5000"]);
> task = FastTimeSeries.TSClassificationSingle(blocks, data);
> backbone = FastTimeSeries.Models.StackedLSTM(1, 16, 10, 2);
> model = FastAI.taskmodel(task, backbone)
Chain(
  StackedLSTMCell(
    Recur(
      LSTMCell(1 => 10),                # 500 parameters
    ),
    Recur(
      LSTMCell(10 => 16),               # 1_760 parameters
    ),
  ),
  identity,
  Dense(16 => 5),                       # 85 parameters
)         # Total: 12 trainable arrays, 2_345 parameters,
          # plus 4 non-trainable, 13_312 parameters, summarysize 62.145 KiB.
```

ToucheSir commented on FastTimeSeries/src/models/StackedLSTM.jl:
```diff
@@ -0,0 +1,50 @@
+of the layers.
+
+"""
+function StackedLSTM(in::Int, out::Integer, hiddensize::Integer, layers::Integer;
+			init=Flux.glorot_uniform)
+	if layers == 1
+		chain = Chain(LSTM(in, out; init=init))
+	elseif layers == 2
+		chain = Chain(LSTM(in, hiddensize; init=init),
+					  LSTM(hiddensize, out; init=init))
+	else
+		chain_vec = [LSTM(in, hiddensize; init=init)]
+		for i = 1:layers - 2
+			push!(chain_vec, LSTM(hiddensize, hiddensize; init=init))
+		end
+		chain = Chain(chain_vec..., LSTM(hiddensize, out; init=init))
+	end
+	return StackedLSTMCell(chain)
+end
```
Something is a little weird here: you're returning a `Cell`, but inside that cell are `Recur`s and not `LSTMCells`. Ideally `StackedLSTMCell` would be structured like the other Flux RNN cells, but if that's not possible I'd recommend renaming to `StackedLSTM` to better represent the model as stateful + containing internal mutation (transitively via `Recur` from `LSTM`).

ToucheSir commented on FastTimeSeries/src/models/StackedLSTM.jl:
```diff
@@ -0,0 +1,50 @@
+#
+# Layers for stacked LSTM (referenced from https://github.com/sdobber/FluxArchitectures.jl)
+
+mutable struct StackedLSTMCell{A}
```
```suggestion
struct StackedLSTMCell{A}
```

ToucheSir commented on FastTimeSeries/src/models/StackedLSTM.jl:
```diff
@@ -0,0 +1,50 @@
+Flux.trainable(m::StackedLSTMCell) = (m.chain)
+
+# Initialize forget gate bias with 1
+function initialize_bias!(l::StackedLSTMCell)
+	for i = 1:length(l.chain)
+		l.chain[i].cell.b .= 1
+	end
+	return nothing
+end
```
This should be done explicitly through the `initb` argument of the LSTM(Cell) constructor if possible.

ToucheSir commented on FastTimeSeries/src/models/StackedLSTM.jl:
```diff
@@ -0,0 +1,50 @@
+end
+
+Flux.@functor StackedLSTMCell 
+Flux.trainable(m::StackedLSTMCell) = (m.chain)
```
```suggestion
```
This is redundant with the `@functor` above.

ToucheSir commented on FastTimeSeries/src/models/StackedLSTM.jl:
```diff
@@ -0,0 +1,50 @@
+end
+
+function (m::StackedLSTMCell)(X)
+	[m.chain(x) for x ∈ X[1:end-1]]
+	return m.chain(X[end])
```
Is this always a sequence-to-one model?

codeboy5 replied:
I think for classification and regression, this will always be a seq-to-one model ?
For some nlp tasks, it would not be.

ToucheSir replied:
Ok, then I think you should be able to rewrite this using `foldl`. It'll be more correct and likely perform better as well. It may also let you drop the `Recur` wrapper for the inner layer stack and use the cells directly instead (i.e. `LSTM -> LSTMCell`).

ToucheSir commented on FastTimeSeries/src/tasks/classification.jl:
```diff
@@ -26,4 +26,33 @@ _tasks["tsclfsingle"] = (
+        row = FastAI.mockblock(task.blocks[1])
+
+        xtrain = FastAI.encodeinput(task, Training(), row);
+        @test length(xtrain[1,:]) == task.blocks.input.obslength
+        @test length(xtrain[:,1]) == task.blocks.input.nfeatures
+
+        @test eltype(xtrain[1,:]) <: Number
```
Is the input expected to be non-contiguous here? If not, I would compare against `size` instead of lengths of slices.

codeboy5 commented:
<img width="348" alt="image" src="https://user-images.githubusercontent.com/40931412/180548868-4a12dc3b-dfc2-48db-a854-52fec9c0cd9b.png">

ToucheSir commented on FastTimeSeries/src/models/RNN.jl:
```diff
@@ -1,3 +1,9 @@
+function tabular2rnn(X::AbstractArray{Float32, 3})
+    X = permutedims(X, (2, 1, 3))
+    X = [X[t, :, :] for t ∈ 1:size(X, 1)]
```
```suggestion
    X = collect(eachslice(X; dims=1))
```
Should be more efficient. But see comments below.

ToucheSir commented on FastTimeSeries/src/models/RNN.jl:
```diff
@@ -24,5 +30,5 @@ function RNNModel(recbackbone,
     dropout = dropout_rate == 0 ? identity : Dropout(dropout_rate)
-    Chain(recbackbone, dropout, finalclassifier)
-end
+    Chain(tabular2rnn, recbackbone, dropout, finalclassifier)
```
I don't think we should be doing this dense to slices transform in the model itself. If you just need something RNN friendly, relying on the built-in support for dense inputs should be enough. The `permutedims` can stay for now, but even that probably shouldn't be in the gradient hot path (it allocates O(n) both ways).

ToucheSir replied:
An alternative for now is to make the data pipeline spit out the vector of arrays. We can then revisit if/when you add models like CNNs which expect dense inputs.

codeboy5 replied:
I think the "ideal" place to do this transform would be inside the training loop. I am not sure how to exactly do that for FastAI.jl. Is there a way ?

Since the second phase would involve using some CNNs, using data pipeline to spit out vector of arrays would not work.

ToucheSir commented on FastTimeSeries/src/models/RNN.jl:
```diff
@@ -1,3 +1,9 @@
+function tabular2rnn(X::AbstractArray{Float32, 3})
```
I think "tabular" is a bit of a misnomer here, but naming is not a priority.

lorenzoh commented:
Since this PR is getting quite long, it may be best to move the newest changes to a new PR and merge the parts that are ready. We could revert the PR to the point where I did the manual merge, merge this PR with the changes up to that point into master, and then open a new PR based on master with the more recent changes, including models and so on. @codeboy5 let me know if you aren’t sure about how to do this 🙂

codeboy5 commented:
> Since this PR is getting quite long, it may be best to move the newest changes to a new PR and merge the parts that are ready. We could revert the PR to the point where I did the manual merge, merge this PR with the changes up to that point into master, and then open a new PR based on master with the more recent changes, including models and so on. @codeboy5 let me know if you aren’t sure about how to do this 🙂

Yeah sure, just have to revert the code to the pr right ?
Will do it.

codeboy5 commented:
@lorenzoh Does this look okay ?

On the last meeting, @darsnack suggested some changes to the model code, I will commit them in the next PR.

lorenzoh commented:
Yeah this looks good! Just to make sure, @darsnack does it make sense to rebase this on master when merging or should we squash and merge?

darsnack commented:
Either is okay. Maybe squash and merge will be less painful?

lorenzoh commented:
I've squashed and merged it. @codeboy5 you should be able to open a new PR for the model code based on master now.

codeboy5 commented:
> I've squashed and merged it. @codeboy5 you should be able to open a new PR for the model code based on master now.

will do thanks . 👍🏻
